File: LeetCode/Python Solutions/Maximum Number of Events That Can Be Attended II/maximum.py
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        # Sort the events by start time
        events.sort(key=lambda x: x[0])

        # Initialize a dictionary to keep track of the maximum value for each number of events
        dp = {}

        # Define a helper function to find the maximum value for each number of events
        def helper(i, k):
            # If we have visited this event before, return the maximum value for this number of events
            if (i, k) in dp:
                return dp[(i, k)]

            # If we have reached the end of the events, return 0
            if i == len(events):
                return 0

            # If we have no more events to attend, return 0
            if k == 0:
                return 0

            # If we have to skip this event, move on to the next event
            if events[i][1] <= events[i][0]:
                return helper(i + 1, k)

            # If we have to attend this event, move on to the next event
            if k == 1:
                return events[i][2] + helper(i + 1, k)

            # If we have to attend this event, move on to the next event
            # Else, we can either attend this event or skip this event
            dp[(i, k)] = max(events[i][2] + helper(i + 1, k), helper(i + 1, k - 1))
            return dp[(i, k)]

        # Return the maximum value for k events
        return helper(0, k)

File: LeetCode/Java Solutions/Maximum Number of Events That Can Be Attended II/maximum.java
import java.util.*;

class Solution {
    // Define a method that takes a 2D array of events and an integer k as inputs
    public int maxValue(int[][] events, int k) {
        // Sort the events array by the end times
        Arrays.sort(events, Comparator.comparingInt(a -> a[1]));

        // Initialize endTimeAndValuePairs to store pairs of [endTime, totalValue]
        List<int[]> endTimeAndValuePairs = new ArrayList<>();
        // The first pair [0, 0] represents the initial state where no event is attended
        endTimeAndValuePairs.add(new int[]{0, 0});

        // Initialize newEndTimeAndValuePairs to store pairs for the next round
        List<int[]> newEndTimeAndValuePairs = new ArrayList<>();
        newEndTimeAndValuePairs.add(new int[]{0, 0});

        // Iterate over each attended event
        for (int attendedEventCount = 0; attendedEventCount < k; attendedEventCount++) {
            for (int[] event : events) {
                // For each event, get start time, end time and value
                int start = event[0], end = event[1], value = event[2];

                // Find the event in endTimeAndValuePairs whose end time is less than the start time of the current event
                int indexOfEvent = Collections.binarySearch(endTimeAndValuePairs, new int[]{start, 0}, Comparator.comparingInt(a -> a[0]));
                if (indexOfEvent < 0) {
                    indexOfEvent = -indexOfEvent - 2;
                }

                // If the total value by attending the current event is larger than the last total value in newEndTimeAndValuePairs
                // Append this new pair to newEndTimeAndValuePairs
                if (indexOfEvent != -1 && endTimeAndValuePairs.get(indexOfEvent)[1] + value > newEndTimeAndValuePairs.get(newEndTimeAndValuePairs.size() - 1)[1]) {
                    newEndTimeAndValuePairs.add(new int[]{end, endTimeAndValuePairs.get(indexOfEvent)[1] + value});
                }
            }

            // Update endTimeAndValuePairs with newEndTimeAndValuePairs for the next round
            endTimeAndValuePairs = new ArrayList<>(newEndTimeAndValuePairs);
            // Reset newEndTimeAndValuePairs for the next round
            newEndTimeAndValuePairs = new ArrayList<>();
            newEndTimeAndValuePairs.add(new int[]{0, 0});
        }

        // Return the maximum value which is the last value in endTimeAndValuePairs
        return endTimeAndValuePairs.get(endTimeAndValuePairs.size() - 1)[1];
    }
}
